### PR TITLE
xDefaultGatewayServer resource added

### PR DESCRIPTION
**Pull request by PlagueHO**
This relates to Issue #18 and issue #22. I also did a small amount of work on Issue #14. Documentation has been updated and example usage has been added. I didn't think the Ensure parameter made sense in the xDefaultGatewayAddress resource as just using it without setting an Address parameter (or leaving it blank) will remove the Default Gateway:
```powershell
xDefaultGatewayAddress RemoveDefaultGateway
{
    InterfaceAlias = 'Ethernet'
    AddressFamily  = 'IPv4'
}
```
Setting the default gateway:
```powershell
xDefaultGatewayAddress SetDefaultGateway
{
    Address = '192.168.0.1'
    InterfaceAlias = 'Ethernet'
    AddressFamily  = 'IPv4'
}
```

Other items:
* Corrected the error messages thrown by the xDNSServerAddress resource when an IP Address format and Address Family mismatches.
* Added check to xIPAddress to ensure that IP address format and Address family matches (copied from xDNSServerAddress resource).
* Added some additional tests to MSFT_xIPAddress.Tests.ps1

### CODE REVIEW

File: DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1 [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    # Use $AddressFamily to select the IPv4 or IPv6 destination prefix
+    $DestinationPrefix = "0.0.0.0/0"
+    if($AddressFamily -eq "IPv6")
+    {
+        $DestinationPrefix = "::/0"
+    }
+    # Get all the default routes
+    $defaultRoutes = Get-NetRoute -InterfaceAlias $InterfaceAlias -AddressFamily `
+        $AddressFamily -ErrorAction Stop | `
+        where { $_.DestinationPrefix -eq $DestinationPrefix }
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1]: Can you change this to Where-Object please?

File: DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1 [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+    ValidateProperties @PSBoundParameters
+}
+
+
+#######################################################################################
+#  Helper function that validates the Gateway Address property. If the switch parameter
+# "Apply" is set, then it will set the properties after a test
+#######################################################################################
+function ValidateProperties
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1]: Please use the verb-noun function name here. Suggesiton: Test-GatewayProperties, or just Test-Properties.

File: DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1 [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        # Use $AddressFamily to select the IPv4 or IPv6 destination prefix
+        $DestinationPrefix = "0.0.0.0/0"
+        if($AddressFamily -eq "IPv6")
+        {
+            $DestinationPrefix = "::/0"
+        }
+        # Get all the default routes
+        $defaultRoutes = Get-NetRoute -InterfaceAlias $InterfaceAlias -AddressFamily `
+            $AddressFamily -ErrorAction Stop | `
+            where { $_.DestinationPrefix -eq $DestinationPrefix }
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.psm1]: Please change this to Where-Object

File: DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.schema.mof [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[ClassVersion("1.0.0"), FriendlyName("xDefaultGatewayAddress")] 
+class MSFT_xDefaultGatewayAddress : OMI_BaseResource
+{
+  [Write] string Address;
+  [Key] string InterfaceAlias;
+  [Key,Write,ValueMap{"IPv4", "IPv6"},Values{"IPv4", "IPv6"}] string AddressFamily;
+};
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDefaultGatewayAddress/MSFT_xDefaultGatewayAddress.schema.mof]: You'll need to add a line break here.

File: DSCResources/MSFT_xIPAddress/MSFT_xIPAddress.psm1 [hunk 5]
```diff
@@ -252,41 +209,32 @@ function ValidateProperties
...
-                    {
-                        $defaultRoutes | Remove-NetRoute -confirm:$false -ErrorAction Stop
-                    }
-
-                    # Remove any IP addresses on the specified interface
-                    if($currentIP)
-                    {
-                        $currentIP | Remove-NetIPAddress -confirm:$false -ErrorAction Stop
-                    }
+                    $defaultRoutes | Remove-NetRoute -confirm:$false -ErrorAction Stop
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xIPAddress/MSFT_xIPAddress.psm1]: Piping is inherantly slower, could we change this to not use a pipe?

Please also captialize all parameters (ie: -Confirm:$false)

File: DSCResources/MSFT_xIPAddress/MSFT_xIPAddress.psm1 [hunk 6]
```diff
@@ -252,41 +209,32 @@ function ValidateProperties
...
-                    if($filterGateway)
-                    {
-                        $filterGateway | Remove-NetRoute -Confirm:$false
-                    }
-
-                    if($filterIP)
-                    {
-                        $filterIP | Remove-NetIPAddress -confirm:$false -ErrorAction Stop
-                    }
+                    $currentIP | Remove-NetIPAddress -confirm:$false -ErrorAction Stop
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xIPAddress/MSFT_xIPAddress.psm1]: Piping is inherantly slower, could we change this to not use a pipe?

Please also captialize all parameters (ie: -Confirm:$false)

**Comment by tysonjhayes** [comment]
This all looks pretty good, some minor changes are needed.

I will say though I have a personal dislike of collapsing test/set logic into the same functions. To me, the way DSC is designed if we are running set then Test has already failed, so why am I running it again? I know some of the design we currently have has a ValidateProperties function but I'd personally like ot see those removed and the appropriate logic moved to the test/set target resource.

Just my personal thoughts once you've cleaned up some of the style stuff I'll get it merged.

**Comment by PlagueHO** [comment]
Great suggestions. I have made them all except for eliminating the ValidateProperties (now named Test-Properties) function. I hadn't thought about that before and had just gone with what seemed to be a common way of doing things (most of the MSFT_xDefaultGatewayAddress resource was copied straight out of MSFT_xIPAddress. I totally agree - this should probably be re-thought to eliminate this double firing of the test code. I'll think on this and keep it in mind. I've also removed the piped cmdlets and turned them into .Where() methods or used the Foreach () function where applicable - hope this was what you were thinking?

Now, as for the AppVeyor failure, I've seen this appear a few times lately but is not consistent and seems to go away if I re-check in the same code:
```
[-] Error occurred in test script 'C:\projects\xnetworking\DscResource.Tests\Meta.Tests.ps1' 32.9s
  Module installation of xDscResourceDesigner failed with exit code 
  at <ScriptBlock>, C:\projects\xnetworking\DscResource.Tests\Meta.Tests.ps1: line 46
```
It is a bit odd because my dev fork the AppVeyor build tests passed fine. So I'll recheck the code in and see if the build works this time.

**Comment by PlagueHO** [comment]
AppVeyor tests passed this time. Go figure.

**Comment by tysonjhayes** [comment]
LGTM. I'm cool with breaking up the logic in seperate pull requests. :+1:

Thanks for the work!

**Comment by PlagueHO** [comment]
No problemo!

**Pull request merged**